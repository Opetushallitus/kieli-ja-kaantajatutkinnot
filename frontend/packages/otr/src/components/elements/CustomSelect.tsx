import {
  FormControl,
  FormHelperText,
  InputLabel,
  MenuItem,
  Select,
  SelectProps,
} from '@mui/material';

type SelectValue = string | number | readonly string[] | undefined;
export interface CustomSelectProps {
  showInputLabel?: boolean;
  values: Map<SelectValue, SelectValue>;
  helperText?: string;
  disableUnderline?: boolean;
  sortByKeys?: boolean;
  showError?: boolean;
}

export const CustomSelect = ({
  showInputLabel,
  id,
  label,
  values,
  helperText,
  showError,
  sortByKeys,
  ...selectOnlyProps
}: CustomSelectProps & SelectProps<string>) => {
  const valuesArray = Array.from(values);
  const valuesToShow = sortByKeys ? valuesArray.sort() : valuesArray;

  const selectProps = { id, label, ...selectOnlyProps };

  return (
    <FormControl fullWidth error={showError}>
      {showInputLabel && <InputLabel id={id}>{label}</InputLabel>}
      <Select autoWidth {...selectProps}>
        {valuesToShow.map(([key, value], index) => (
          <MenuItem key={index} value={value}>
            {key}
          </MenuItem>
        ))}
      </Select>
      {showError && <FormHelperText>{helperText}</FormHelperText>}
    </FormControl>
  );
};
