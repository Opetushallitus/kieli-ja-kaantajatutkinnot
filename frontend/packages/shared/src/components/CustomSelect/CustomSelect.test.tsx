import renderer from 'react-test-renderer';

import { CustomSelect } from './CustomSelect';

describe('CustomSelect', () => {
  it('should render correctly', () => {
    const values = new Map<string, string>([
      ['Option A', 'A'],
      ['Option B', 'B'],
      ['Option C', 'C'],
    ]);
    const tree = renderer
      .create(<CustomSelect values={values} value="Option B" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
