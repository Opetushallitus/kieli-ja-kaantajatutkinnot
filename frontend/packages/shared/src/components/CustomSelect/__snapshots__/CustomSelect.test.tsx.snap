// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomSelect should render correctly 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    onClick={[Function]}
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="notranslate"
      >
        ​
      </span>
    </div>
    <input
      aria-hidden={true}
      aria-invalid={false}
      className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
      disabled={false}
      onAnimationStart={[Function]}
      onChange={[Function]}
      required={false}
      tabIndex={-1}
      value="Option B"
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
      data-testid="ArrowDropDownIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-ihdtdm"
      >
        <span
          className="notranslate"
        >
          ​
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
